osx/KBKit/KBKit/System/KBTask.m

```diff
@@ -9,34 +9,65 @@
 #import "KBTask.h"
 
 #import "KBDefines.h"
+#include <signal.h>
 
 @implementation KBTask
 
-+ (void)execute:(NSString *)command args:(NSArray *)args completion:(void (^)(NSError *error, NSData *outData, NSData *errData))completion {
++ (void)execute:(NSString *)command args:(NSArray *)args timeout:(NSTimeInterval)timeout completion:(void (^)(NSError *error, NSData *outData, NSData *errData))completion {
   NSTask *task = [[NSTask alloc] init];
   task.launchPath = command;
   task.arguments = args;
+  [task setStandardInput:[NSPipe pipe]];
   NSPipe *outPipe = [NSPipe pipe];
   [task setStandardOutput:outPipe];
   NSPipe *errPipe = [NSPipe pipe];
   [task setStandardError:errPipe];
-  task.terminationHandler = ^(NSTask *t) {
-    //DDLogDebug(@"Task: \"%@ %@\" (%@)", command, [args componentsJoinedByString:@" "], @(t.terminationStatus));
-    NSFileHandle *outRead = [outPipe fileHandleForReading];
-    NSData *outData = [outRead readDataToEndOfFile];
-    NSFileHandle *errRead = [errPipe fileHandleForReading];
-    NSData *errData = [errRead readDataToEndOfFile];
-    dispatch_async(dispatch_get_main_queue(), ^{
-      // TODO Check termination status and complete with error if > 0
-      DDLogDebug(@"Task (done): %@, %@", @(t.terminationStatus), @(t.terminationReason));
-      if ([outData length] > 0) {
-        DDLogDebug(@"Task (out): %@", [[NSString alloc] initWithData:outData encoding:NSUTF8StringEncoding]);
-      }
-      if ([errData length] > 0) {
-        DDLogDebug(@"Task (err): %@", [[NSString alloc] initWithData:errData encoding:NSUTF8StringEncoding]);
+
+  dispatch_queue_t taskQueue = dispatch_queue_create("taskQueue", NULL);
+
+  __block BOOL replied = NO;
+  __block BOOL timedOut = NO;
+  if (timeout > 0) {
+    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, timeout * NSEC_PER_SEC), taskQueue, ^{
+      if (!replied) {
+        timedOut = YES;
+
+        // Dump golang stack to log
+        kill(task.processIdentifier, SIGABRT);
+
+        [self readOutPipe:outPipe errPipe:errPipe completion:^(NSData *outData, NSData *errData) {
+          NSString *taskDesc = NSStringWithFormat(@"%@ %@", command, [args join:@" "]);
+          DDLogError(@"Task timed out: %@", taskDesc);
+
+          dispatch_async(dispatch_get_main_queue(), ^{
+            completion(KBMakeError(KBErrorCodeTimeout, @"Task timed out: %@", taskDesc), nil, nil);
+          });
+
+          // Terminate/cleanup task after timeout
+          @try {
+            [task terminate];
+          } @catch(NSException *e) {
+            DDLogDebug(@"Task error in terminate after timeout: %@", e);
+          }
+        }];
       }
+    });
+  }
 
-      completion(nil, outData, errData);
+  task.terminationHandler = ^(NSTask *t) {
+    dispatch_async(taskQueue, ^{
```

Comment:
Discussed IRL: probably can race and make a regular terminate look like a timeout, but this is of marginal impact because a moment later and it would have timed out anyway.